Documentation/workflow/DevelopmentTips.md

```diff
@@ -1,37 +1,292 @@
+     W/monodroid( 2796): Creating public update directory: `/data/data/Mono.Android_Tests/files/.__override__`
 
 When the app needs to resolve native libraries and assemblies, it will look
 for those files within the update directory *first*. This includes the Mono
 runtime library and BCL assemblies.
 
 Note that the update directory is *per-app*. The above mentioned `Mono.Android_Tests`
 directory is created when running the
-[`Mono.Android-Tests.csproj`](../src/Mono.Android/Test/Mono.Android-Tests.csproj)
+[`Mono.Android-Tests.csproj`](../../src/Mono.Android/Test/Mono.Android-Tests.csproj)
 unit tests.
 
 The update directory is not used in `Release` configuration builds.
 (Note: `Release` configuration for the *app itself*, not for xamarin-android.)
 
-For example, if you're working on a mono/x86 bug and need to quickly update
-the app on the device to test `libmonosgen-2.0.so` changes:
+Keep in mind that only the app that owns the update directory has permission to
+write to it, so on a normal non-rooted device or emulator, you'll need to use
+`adb shell run-as` to add new files.  For example, if you're working on a
+mono/x86 bug and need to quickly update the app on the device to test
+`libmonosgen-2.0.so` changes:
 
-	$ make -C src/mono-runtimes/obj/Debug/x86 && \
-	  adb push src/mono-runtimes/obj/Debug/x86/mono/mini/.libs/libmonosgen-2.0.so \
-	    /data/data/Mono.Android_Tests/files/.__override__
+    $ make -C src/mono-runtimes/obj/Debug/x86 && \
+      adb push src/mono-runtimes/obj/Debug/x86/mono/mini/.libs/libmonosgen-2.0.so \
+        /data/local/tmp/ && \
+      adb shell run-as Mono.Android_Tests cp /data/local/tmp/libmonosgen-2.0.so \
+        /data/data/Mono.Android_Tests/files/.__override__/
 
 Alternatively, if you're working on an `mscorlib.dll` bug:
 
-	$ make -C external/mono/mcs/class/corlib PROFILE=monodroid && \
-	  adb push external/mono/mcs/class/lib/monodroid/mscorlib.dll \
-	    /data/data/Mono.Android_Tests/files/.__override__
+    $ make -C external/mono/mcs/class/corlib PROFILE=monodroid && \
+      adb push external/mono/mcs/class/lib/monodroid/mscorlib.dll \
+        /data/local/tmp/ && \
+      adb shell run-as Mono.Android_Tests cp /data/local/tmp/mscorlib.dll \
+        /data/data/Mono.Android_Tests/files/.__override__/
 
+On some devices, the `run-as` command might not have permission to read the
+files in `/data/local/tmp/`.  In that case, you can use a `cat` command that
+*pipes to* the `run-as` command:
+
+    $ make -C external/mono/mcs/class/corlib PROFILE=monodroid && \
+      adb push external/mono/mcs/class/lib/monodroid/mscorlib.dll \
+        /data/local/tmp/ && \
+      adb shell "cat /data/local/tmp/mscorlib.dll | \
+        run-as Mono.Android_Tests sh -c \
+        'cat > /data/data/Mono.Android_Tests/files/.__override__/mscorlib.dll'"
+
+## Attaching LLDB using Android Studio on Windows or macOS
```

Comment:
This should be promoted to a `#` section, and should come *last*, for two reasons:

 1. It's longest, and thus it's presence first "buries" the other two sections, making them harder to see (as they're below the proverbial "fold")
 2. It gives the impression that Android Studio is the "best" way to debug here.

Even if (2) is true, that's not a message we want to promote. :-)